[DOC REVIEW: PR 1512] Running in production over k8s (#1522)
nuclio/nuclio: docs/setup/k8s/running-in-production-k8s.md

# Running Nuclio Over Kubernetes in Production

After familiarizing yourself with Nuclio and [deploying it over Kubernetes](/docs/setup/k8s/getting-started-k8s.md), you might find yourself in need of more information pertaining to running Nuclio in production.
Nuclio is integrated, for example, within the [Iguazio Data Science Platform](https://www.iguazio.com), which is used extensively in production, both by Iguazio and its customers, running various workloads.
This document describes advanced configuration options and best-practice guidelines for using Nuclio in a production environment.

#### In this document

- [Selecting the deployment method](#selecting-the-deployment-method)
- [Freezing a qualified version](#version-freezing)
- [Multi-Tenancy](#multi-tenancy)
- [Air-gapped deployment](#air-gapped-deployment)
- [Using Kaniko as an image builder](#kaniko-image-builder)

<a id="Selecting the deployment method"></a>

> ## Selecting the deployment method

Author: OK, I changed it to "The preferred deployment method" — see commit 9981089.

> You can use any of the deployment (installation) methods described in the Nuclio [Setup](/docs/setup/) documentation also in production.

Author: OK, but the original phrasing of "any of the available methods" is very vague and unclear. (That's why I wanted to clarify with you yesterday which methods you were referring to). => Can you suggest an alternative?

> You can use any of the deployment (installation) methods described in the Nuclio [Setup](/docs/setup/) documentation also in production.

Reviewer: @Sharon-iguazio  - yes, keep it vague :)

> You can use any of the deployment (installation) methods described in the Nuclio [Setup](/docs/setup/) documentation also in production.

Author: OK, I rephrased as follows — see commit 7080ede:
> There are several alternatives to deploying (installing) Nuclio in production, but the recommended method is by using [Helm charts](/hack/k8s/helm/nuclio/).
> This is the currently the preferred deployment method at Iguazio as it's the most tightly maintained, it's best suited for "heavy lifting" over Kubernetes, and it's often used to roll out new production-oriented features

> You can use any of the deployment (installation) methods described in the Nuclio [Setup](/docs/setup/) documentation also in production.

Reviewer: `the currently the preferred` -> `currently the preferred

> You can use any of the deployment (installation) methods described in the Nuclio [Setup](/docs/setup/) documentation also in production.

Author: Ugh! Fixed — see commit 23b01a4.

- Set `dashboard.baseImagePullPolicy` to `Never`.
- Set `registry.pushPullUrl` to a registry URL that's reachable from your system.
- <a id="air-gapped-envir-processor-n-onbuild-images"></a>Ensure that the processor and "onbuild" images are accessible to the dashboard in your environment, as they're required for the build process (either by `docker build` or [Kaniko](#using-kaniko-as-an-image-builder)).
  You can achieve this using either of the following methods:

  - Make the images available to the Kubernetes Docker daemon.
  - Preload the images to a registry that's accessible to your system, to allow pulling the images from the registry.
    When using this method, set `registy.defaultBaseRegistryURL` to the URL of an accessible local registry that contains the preloaded images (thus overriding the default location of `quay.io/nuclio`, which isn't accessible in air-gapped environments).
    <br/><br/>
    > **Note:** To save yourself some work, you can use the [prebaked Nuclio registry](https://github.com/nuclio/prebaked-registry), either as-is or as a reference for creating your own local registry with preloaded images.

> - To use a Nuclio templates library (optional), package the templates into an archive, and set `dashboard.templatesArchiveAddress` to a local address for serving the archive, which is accessible to your system.

Reviewer: ```suggestion
- To use the Nuclio templates library (optional), package the templates into an archive and serve them via a local server of your doing on some address accessible to the system. Set `dashboard.templatesArchiveAddress` to the address of that local server.
```

Author: I changed it to the following — see commit f00f5c2:
> To use the Nuclio templates library (optional), package the templates into an archive; serve the templates archive via a local server whose address is accessible to your system; and set `dashboard.templatesArchiveAddress` to the address of this local server

        ./hack/k8s/helm/nuclio/
    ```

<a id="multi-tenancy"></a>
## Multi-Tenancy

Implementation of multi-tenancy can be done in many different ways and to various degrees.
The experience of the Nuclio team has lead to the adoption of the Kubernetes approach of tenant isolation using namespaces.
Note:

- To achieve tenant separation for various Nuclio projects and functions, and to avoid cross-tenant contamination and resource races, a fully functioning Nuclio deployment is used in each namespace and the Nuclio controller is configured to be namespaced.
  This means that the controller handles Nuclio resources (functions, function events, and projects) only within its own namespace.
  This is supported by using the `controller.namespace` [Helm values](/hack/k8s/helm/nuclio/values.yaml) configuration.

> - To provide ample separation at the level of the container image, it's highly recommended that the Nuclio deployments of multiple tenants either don't share container registries, or that they don't share a tenant when using a multi-tenant registry (such as `docker.io` or `quay.io`).

Author: OK, I changed "container image" to "container registry" — see commit 3ccdd95; (we don't use "container image registry" anywhere).

<a id="version-freezing"></a>
## Freezing a qualified version

When working in production, you need reproducibility and consistency.
It's therefore recommended that you don't use the latest stable version, but rather qualify a specific Nuclio version and "freeze" it in your configuration.
Stick with this version until you qualify a newer version for your system.

> Because Nuclio adheres to backwards-compatibility standards between patch versions, and even minor version updates don't typically break major functionality, the process of qualifying a newer Nuclio version should typically be short and easy.

Reviewer: `typically` used twice

Author: Nice catch. I changed the second instance to "generally" — see commit bf210de.

The following guidelines refer to more advanced use cases and are based on the assumption that you can handle the related DevOps tasks.
Note that such implementations can get a bit tricky; to access a fully-managed, air-gap friendly, "batteries-included", Nuclio deployment, which also offers plenty of other tools and features, check out the enterprise-grade [Iguazio Data Science Platform](https://www.iguazio.com/platform/).
If you select to handle the implementation yourself, follow these guidelines; the referenced configurations are all [Helm values](/hack/k8s/helm/nuclio/values.yaml):

- Set `*.image.repository` and `*.image.tag` to [freeze a qualified version](#version-freezing), and ensure that the configured images are accessible to the Kubernetes deployment.
- Set `*.image.pullPolicy` to `Never` or to `IfNotPresent` to ensure that Kubernetes doesn't try to fetch the images from the web.
- Set `offline` to `true` to put Nuclio in "offline" mode.
- Set `dashboard.baseImagePullPolicy` to `Never`.
- Set `registry.pushPullUrl` to a registry URL that's reachable from your system.

> - To use a Nuclio templates library (optional), set `dashboard.templatesArchiveAddress` to a location for serving the templates.

Reviewer: > "location for serving" sounds a bit off
Maybe:
```
To use a Nuclio templates library (optional), you need to package the templates yourself and ensure that they're served locally in a way that's accessible to your system, and set `dashboard.templatesArchiveAddress` to point to where they are served.
```

Author: I changed it to the following (I took my cue also from the value name) — see commit dfd8747:
> - To use a Nuclio templates library (optional), package the templates into an archive that's served locally and accessible to your system, and set `dashboard.templatesArchiveAddress` to the archive address

Reviewer: That is a bit inaccurate as "archive" doesn't have an address. it is served on some server that serves it on some address (we're talking http serving here)

  Note that you need to package the templates yourself and ensure that they're served locally in a way that's accessible to your system.

> <a id="air-gapped-envir-processor-n-on-build-images"></a>In addition, the processor and on-build images must also be accessible to the dashboard in your environment, as they're required for the build process (either by `docker build` or [Kaniko](#using-kaniko-as-an-image-builder)).

Reviewer: This needs to follow line 101 (`registry.pushPullUrl`), `templatesArchiveAddress` can come afterwards

Author: The reason that I moved it out of the bulleted list was that I tried to keep the list phrasing parallel ("Set xxx ..."), as good writing conventions dictate; that's also why I initially phrased the `dashboard.templatesArchiveAddress` bullet so that "set" comes as soon as possible. But the processor and on-line images bullet is different because it doesn't refer only to values configurations. => Are you adamant that we can't keep the current structure?

Reviewer: Yes, since this follows both in content and in importance. It is in close affinity to setting `pushPullUrl` and both have to do with making images available to your cluster (the first referring to where result images will be pushed and this refers to where the source images (onbuild, base) are to be pulled from.
The templatesArchive are a different kind of beast and are also less critical to the core functionality

Author: OK, I moved it back to be the bulleted list and slightly rephrased the opening sentence accordingly — see commit ffde717.

> <a id="air-gapped-envir-processor-n-on-build-images"></a>In addition, the processor and on-build images must also be accessible to the dashboard in your environment, as they're required for the build process (either by `docker build` or [Kaniko](#using-kaniko-as-an-image-builder)).

Reviewer: ```suggestion
<a id="air-gapped-envir-processor-n-on-build-images"></a>In addition, the processor and onbuild images must also be accessible to the dashboard in your environment, as they're required for the build process (either by `docker build` or [Kaniko](#using-kaniko-as-an-image-builder)).
```
`onbuild` is a term from the code and is one word in the name of these images as well

Author: I changed it to _"onbuild"_ (with the quotation marks) in the two locations in which it was used in the text, as I saw this is what we did in other doc locations and in the UI and this is also what's done in some of the Docker and k8s docs — see commit 424d83c.

Reviewer: np

You can achieve this using either of the following methods:

- Make the images available to the Kubernetes Docker daemon.
- Preload the images to a registry that's accessible to your system, to allow pulling the images from the registry.

>   When using this method, set `registy.defaultBaseRegistryURL` to the base URL to be used for searching your registry for the images (instead of the default location of `quay.io/nuclio`, which isn't accessible in air-gapped environments).

Reviewer: ```suggestion
  When using this method, set `registy.defaultBaseRegistryURL` to the URL of the registry used for the base and onbuild images (instead of the default location of `quay.io/nuclio`, which isn't accessible in air-gapped environments).
```

Author: I thought "base" in `registy.defaultBaseRegistryURL` refers to the URL, not the images, which seemed to align with the "searching ..." text, which I took from the original doc version. I don't think we should say "base and onbuild images" because we don't refer to "base images" anywhere else, we refer to "processor and on-build images". =>
I've now changed it to the following — see commit bdbfe19. I used "the URL of an accessible local registry that contains the preloaded images" (similar to to the related Kaniko doc, which I've now edited to remove "base" — per your separate request — but without the image types, which are already listed in the preceding text); I also edited the text in parentheses for more accuracy:
> When using this method, set `registy.defaultBaseRegistryURL` to the URL of an accessible local registry that contains the preloaded images (thus overriding the default location of `quay.io/nuclio`, which isn't accessible in air-gapped environments)

Reviewer: > "we don't refer to "base images" anywhere else, we refer to "processor and on-build images" - base images are a crucial part of building a nuclio function container image and we'll need to refer to them in this doc.
That being said, I've unearthed what looks like a bug in how we refer to those today via this specific value `defaultBaseRegistryURL`, which causes this to only effect onbuild images and not base images.
base images can be controlled via a different "knob" I'll be adding shortly. Naming is not ideal here...
So, let's go with: `to the URL of the registry used for the onbuild images` for now, and i'll add a dedicated bulletin for base images once i'm done fixing / testing

Author: Currently, the doc refers to 'the URL of an accessible local registry that contains the preloaded processor and "onbuild" images'. As I understand from you, we shouldn't add the base-images reference at this stage, as it requires a new configuration. I wasn't sure whether you want to keep the reference to the processor images or refer only to the "onbuild" images at this stage, so I kept it for now (as it was used in the original text).
(BTW, I checked and our current terminology in the function configuration reference is '"onbuild" image, "base container image", and "function processor image".)

Reviewer: Keep the reference to the processor images 👍

Author: Already done ... .

    .
```

This is rather straightforward; however, note the following:

- When running in an [air-gapped environment](#air-gapped-deployment), Kaniko's executor image must also be available to your Kubernetes cluster.
- Kaniko requires that you work with a registry to which push the resulting function images.
  It doesn't support accessing images on the host Docker daemon.

>   Therefore, you must set `registry.pushPullUrl` to the URL of the registry to which Kaniko should push the resulting images, and in air-gapped environments, you must also set `registry.defaultBaseRegistryURL` to the base URL of an accessible local registry that contains the preloaded processor and on-build images (see [Air-gapped deployment](#air-gapped-envir-processor-n-on-build-images)).

Reviewer: ```suggestion
  Therefore, you must set `registry.pushPullUrl` to the URL of the registry to which Kaniko should push the resulting images, and in air-gapped environments, you must also set `registry.defaultBaseRegistryURL` to the URL of an accessible local registry that contains the preloaded processor and on-build images (see [Air-gapped deployment](#air-gapped-envir-processor-n-on-build-images)).
```

Author: Done — see commit bdbfe19.
